### PR TITLE
Bidirectional parsing works.

### PR DESCRIPTION
**Pull request by SerenadeX**
Solves #2

### CODE REVIEW

File: Example/Tests/NSAttributedString+MarkdownTests.swift [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+//  Copyright © 2016 CocoaPods. All rights reserved.
+//
+
+import Foundation
+import UIKit
+import XCTest
+import TSSwiftMarkdownParser
+
+class NSAttributedStringTests: XCTestCase {
+    func testTextWithNoFormatting() {
```
Comment (brandenr) [inline Example/Tests/NSAttributedString+MarkdownTests.swift:16]: There should be an empty line after class

File: Example/Tests/NSAttributedString+MarkdownTests.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        let attributedString = NSAttributedString(string: "This is text that looks like it should be _italicized_ or **bolded**, but really isn't because it wasn't written in markdown.")
+        let escapedMarkdown = "This is text that looks like it should be \\_italicized\\_ or \\*\\*bolded\\*\\*, but really isn't because it wasn't written in markdown\\."
+        XCTAssertEqual(attributedString.markdownString(), escapedMarkdown)
+    }
+    
+    func assertRoundTripFromMarkdownString(markdownString: String) {
+        let attributedString = TSSwiftMarkdownParser(withDefaultParsing: true).attributedStringFromMarkdown(markdownString)
+        XCTAssertEqual(attributedString!.markdownString(), markdownString)
+    }
+}
```
Comment (brandenr) [inline Example/Tests/NSAttributedString+MarkdownTests.swift]: There should be an empty line before and after this }

File: Example/Tests/TSSwiftMarkdownParserTests.swift [hunk 3]
```diff
@@ -205,9 +205,10 @@ class TSSwiftMarkdownParserTests: XCTestCase {
             attributedString.replaceCharactersInRange(range, withString: "    • ")
         }, textFormattingBlock: nil)
         
-        parser.addStrongParsingWithFormattingBlock { attributedString, range in
+
```
Comment (brandenr) [inline Example/Tests/TSSwiftMarkdownParserTests.swift]: Undo this change. You don't need the ()s and you don't need the second empty line

File: Pod/Classes/NSAttributedString+Markdown.swift [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                }
+                
+                markdownString += formatString
+                
+                stringHasBoldEnabled = traits.contains(.TraitBold)
+                stringHasItalicEnabled = traits.contains(.TraitItalic)
+            }
+            
+            let preprocessedString = (self.string as NSString).substringWithRange(range)
+            let processedString = preprocessedString.characters.reduce("") { resultString, character -> String in
```
Comment (brandenr) [inline Pod/Classes/NSAttributedString+Markdown.swift]: You don't need the `-> String`

File: Pod/Classes/NSAttributedString+Markdown.swift [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                        stringToAppend = "\(character)"
+                    }
+                }
+                
+                previousCharacter = character
+                return "\(resultString)\(stringToAppend)"
+            }
+            
+            markdownString += processedString
+
```
Comment (brandenr) [inline Pod/Classes/NSAttributedString+Markdown.swift]: Remove extra empty line

File: Pod/Classes/NSAttributedString+Markdown.swift [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                    }
+                    if previousCharacter == nil { numberedListIsFirstLine = true }
+                    
+                    stringToAppend = "\(character)"
+                case bulletCharacter:
+                    characterOnBulletedListLine = true
+                    stringToAppend = "+ \(previousCharacter != nil ? String(closingString.characters.reverse()) : markdownString)"
+                    if previousCharacter == nil { markdownString = "" }
+                case ".":
+
```
Comment (brandenr) [inline Pod/Classes/NSAttributedString+Markdown.swift]: Remove extra empty line

File: Pod/Classes/NSAttributedString+Markdown.swift [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                case "\n", "\u{2028}":
+                    stringToAppend = "\(closingString)\(character)"
+                    if !characterOnBulletedListLine && !characterOnNumberedListLine {
+                        stringToAppend += String(closingString.characters.reverse())
+                    }
+                case "1", "2", "3", "4", "5", "6", "7", "8", "9", "0":
+                    if previousCharacter == "\n" || previousCharacter == nil {
+                        openedNumberedListStarter = true
+                    }
+                    if previousCharacter == nil { numberedListIsFirstLine = true }
```
Comment (brandenr) [inline Pod/Classes/NSAttributedString+Markdown.swift]: This shouldn't be all on one line

File: Pod/Classes/NSAttributedString+Markdown.swift [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                    
+                    stringToAppend = "\(character)"
+                case bulletCharacter:
+                    characterOnBulletedListLine = true
+                    stringToAppend = "+ \(previousCharacter != nil ? String(closingString.characters.reverse()) : markdownString)"
+                    if previousCharacter == nil { markdownString = "" }
```
Comment (brandenr) [inline Pod/Classes/NSAttributedString+Markdown.swift]: Same

File: Pod/Classes/TSSwiftMarkdownParser.swift [hunk 9]
```diff
@@ -67,6 +67,8 @@ public class TSSwiftMarkdownParser: TSBaseParser {
         attributedString.addAttributes(newAttributes, range: range)
     }
     
+
```
Comment (brandenr) [inline Pod/Classes/TSSwiftMarkdownParser.swift]: Undo adding these two extra empty lines

File: Pod/Classes/TSSwiftMarkdownParser.swift [hunk 10]
```diff
@@ -143,20 +145,22 @@ public class TSSwiftMarkdownParser: TSBaseParser {
             }
             
             addStrongParsingWithFormattingBlock { attributedString, range in
-                if let font = attributedString.attributesAtIndex(range.location, longestEffectiveRange: nil, inRange: range)[NSFontAttributeName] as? UIFont,
-                    italicFont = self.emphasisAttributes[NSFontAttributeName] as? UIFont where font == italicFont {
-                    attributedString.addAttributes(self.strongAndEmphasisAttributes, range: range)
-                } else {
-                    attributedString.addAttributes(self.strongAttributes, range: range)
-                }
+                    attributedString.enumerateAttributesInRange(range, options: []) { attributes, range, _ in
```
Comment (brandenr) [inline Pod/Classes/TSSwiftMarkdownParser.swift]: This shouldn't have been indented

File: Pod/Classes/NSAttributedString+Markdown.swift (hunk 7 above)
Comment (SerenadeX) [inline Pod/Classes/NSAttributedString+Markdown.swift]: I want to argue that point.  I think if we have any block with a small one liner it should be ok.  We do it with `guard` all the time

**Comment by brandenr** [comment]
.

File: Pod/Classes/NSAttributedString+Markdown.swift (hunk 7 above)
Comment (brandenr) [inline Pod/Classes/NSAttributedString+Markdown.swift]: Go ahead and argue it.

Comment (SerenadeX) [inline Pod/Classes/NSAttributedString+Markdown.swift]: I just did.  If there is a one line statement inside a block, I think it makes more sense and looks better to keep the logic check and the block on one line.  We do it with `guard` and our `guard` statements can get pretty long.

Comment (brandenr) [inline Pod/Classes/NSAttributedString+Markdown.swift]: Well, I disagree, since guard and if have different meanings. Guard is used more as an assert. It's to say that there isn't any point continuing on if that stuff isn't true. So, if you want to argue the point, you're going to do better in Slack than here.

Comment (brandenr) [inline Pod/Classes/NSAttributedString+Markdown.swift]: Also, we don't do this anywhere else in the code and if you want the style to be allowed, you'll need Stephan and/or majority vote to allow it.
However, this is your own repo, so you could just ignore me on this and leave it.

File: Example/Tests/NSAttributedString+MarkdownTests.swift [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        assertRoundTripFromMarkdownString("Some text that has _italics and **bold**_")
+    }
+    
+    func testMultipleFormatting() {
+        let strings = [
+            "**Markdown _double_**_ something_",
+            "_italic **double**_** bold**"
+        ]
+        
+        for string in strings {
```
Comment (brandenr) [inline Example/Tests/NSAttributedString+MarkdownTests.swift]: Just because our one line discussion...
strings.forEach { assertRoundTripFromMarkdownString($0) }

File: Pod/Classes/NSAttributedString+Markdown.swift [hunk 12]
```diff
@@ -0,0 +1,136 @@
+import Foundation
+import UIKit
+
+public extension NSAttributedString {
+    
+    public func markdownString() -> String {
+
```
Comment (brandenr) [inline Pod/Classes/NSAttributedString+Markdown.swift]: Extra empty line

**Comment by brandenr** [comment]
👍

**Comment by brandenr** [comment]
👍
